### PR TITLE
dash/20756-edit-mode-change-event

### LINKED ISSUE #20756: Dashboard layout and configuration change event
# Description of the feature
Currently the way edit mode works is that the user needs to commit changes.
If we set edit mode, we need to close the edit mode to export the state of the dashboard.
When the user edits or configures a widget, the also need to confirm the change to see the impact of their edit.

I think it would be a very good user experience if we were able to listen to change events so that we can persist them immediately.
Say a user resizes a widget, it would trigger an event where we could get the JSON representation of the dashboard.
Same if we drag a new widget, remove something or reconfigure something.

Finally, an extra layer to this event could also be to the chart configuration sidebar that instead of confirming changes we will be able to see the impact immediately on change.

# Library related to the feature
Dashboards

### PR DESCRIPTION
**Pull request by vazonik**
Added [edit mode events](https://www.highcharts.com/docs/dashboards/edit-mode#edit-mode-events). Changed the way components are updated in edit mode. See #20756.

---
Closes #20756 

TO DO:
- [x] Change the moment of component update in the edit mode accordion.
- [x] Add edit-mode events called when updating component options.
- [x] Add edit-mode events called when editing the layout.
- [x] Add a modal popup confirming whether to cancel component editing when clicked out of the accordion.
- [x] Add regression tests.
- [x] Update docs.

---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1206750034786913

### CODE REVIEW

**Review by jomar-honsi — APPROVED** [review]

File: ts/Dashboards/EditMode/AccordionMenu.ts [hunk 1]
```diff
@@ -288,6 +336,114 @@ class AccordionMenu {
...
+     * @fires EditMode#componentChanged
+     */
+    private async confirmChanges(): Promise<void> {
+        const component = this.component;
+        if (!component) {
+            return;
+        }
+
+        if (Object.keys(this.chartOptionsJSON).length) {
+            await component.update({
```
Comment (karolkolodziej) [inline ts/Dashboards/EditMode/AccordionMenu.ts:364]: This is only true for the Highcharts component right? Wouldn't it be possible to improve the typing?

**Review by karolkolodziej — CHANGES_REQUESTED** [review]
1. Maskiis missing steps: edit component -> click outside of the sidebar -> click cancel 
2. IMO clicking outside of the dashboard's container should not trigger the popup

File: test/cypress/dashboards/integration/EditMode/sidebar.cy.js [hunk 2]
```diff
@@ -19,13 +16,42 @@ describe('Edit Mode sidebar', () => {
...
-        // Open sidebar
-        cy.get('#dashboard-col-1').click();
-        cy.get('.highcharts-dashboards-edit-toolbar-cell').children().eq(0).click();
         cy.get('.highcharts-dashboards-edit-accordion-header-btn').eq(1).click();
 
         // Assess
         cy.get('.highcharts-dashboards-edit-accordion-content input').should('have.value', '')
     });
+
+    it('Should be able to confirm or cancel changes, #20756.', () => {
```
Comment (karolkolodziej) [inline test/cypress/dashboards/integration/EditMode/sidebar.cy.js:27]: Can you please add test coverage for the visibility of the mask throughout the whole process?

**Review by karolkolodziej — APPROVED** [review]

**Review by jomar-honsi — APPROVED** [review]

**Review by jomar-honsi — APPROVED** [review]

**Review by sebastianbochan — APPROVED** [review]

**Review by stitot — COMMENTED** [review]
Nice work! Are docs unclear on the fact that not all changes will trigger events, such as component (not chart) options such as title? Possible to add support for these as well?

PS. Really nice(!) that components are updated when fields are unfocused!

**Comment by vazonik** [comment]
> Are docs unclear on the fact that not all changes will trigger events, such as component (not chart) options such as title? Possible to add support for these as well?

In what case did you observe this, @stitot? My point was that events are always emitted when confirming or cancelling changes, regardless of the option type. I can't reproduce this, it seems to work here too.
image

**Comment by stitot** [comment]
Hm, it work fine now, @vazonik . When I tried last time only chart options, not component, triggered the event. All good!

**Pull request merged**